### PR TITLE
call Backend.matrix_type to get BGEN locus type

### PR DESCRIPTION
**Pull request by cseed**
clean up on Scala side
there is a full suite of logistic regression tests on the Python side (in methods/test_statgen.py)

### CODE REVIEW

**Comment by cseed** [comment]
Just a heads up this is failing in `TestUtilsSuite.testDataProviders` but I can't reproduce it locally.  I'll let you know once I figure out what's going on.

**Comment by cseed** [comment]
OK, should be fixed now.

File: hail/src/test/scala/is/hail/TestUtils.scala [hunk 1]
```diff
@@ -550,43 +550,6 @@ object TestUtils {
     new MatrixTable(hc, MatrixRead(reader.fullType, dropSamples, false, reader))
   }
 
-  def importBgens(hc: HailContext,
-    files: Seq[String],
-    sampleFile: Option[String] = None,
-    includeGT: Boolean = true,
```
Comment (akotlar) [inline hail/src/test/scala/is/hail/TestUtils.scala]: I like the new handling of includeGT, GP, etc; seems neater to store in (mr: MatrixRead).typ.entryType

File: hail/python/hail/methods/impex.py [hunk 2]
```diff
@@ -953,16 +953,15 @@ def import_bgen(path,
...
     row_set = set(_row_fields)
 
-    # have to get reference genome from the index files!
-    reference_genome = Env.hail().io.bgen.LoadBgen.getReferenceGenome(Env.hc()._jhc.hadoopConf(), wrap_to_list(path), java_index_file_map)
-    lt = tlocus(reference_genome) if reference_genome else tstruct(contig=tstr, position=tint32)
-
     if variants is not None:
+        mt_type = Env.backend().matrix_type(
+            MatrixRead(MatrixBGENReader(path, sample_file, index_file_map, n_partitions, block_size, None)))
+        lt = mt_type.row_type['locus']
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/python/hail/methods/impex.py]: `variants` is a set of variants to restrict to when importing.  Therefore, the type of the variants have to match the type of locus in the bgen file.  So we grab the type of the bgen file by pretending we're importing (without restricting) and get the type.  In this sense, it must be "consistent".  That's all that `LoadBgen.getReferenceGenome` is doing.  Note, `getReferenceGenome` is still called by the `MatrixBGENReader` constructor in Scala and used to construct locus type, see for example, `TLocus.schemaFromRG`.

**Comment by cseed** [comment]
Great questions.

The larger change here is that we're trying to decouple the Python front end from Scala/Java.  Therefore, instead of calling directly into Python (e.g. getReferenceGenome), we're constructing a MatrixRead IR, which can either be passed to a service (ServiceBackend) or passed to Java as we did before (SparkBackend).  I should also remark, py4j is incredibly slow, so batching calls to Python greatly improves performance in the current setup.  So, for example, building an IR in  Python, serializing it and parsing on Java is much faster than a series of py4j calls that builds the corresponding objects on the Java side one at a time.

Some functions which used to be called from Python like importBgens are no longer used, but are still used by the Scala tests.  Those got moved to TestUtils and are being phased out.

**Comment by akotlar** [comment]
Thanks!

**Pull request merged**